cirq-core/cirq/protocols/kraus_protocol.py

```diff
@@ -151,36 +155,55 @@ def kraus(
             method returned NotImplemented) and also no default value was
             specified.
     """
-    channel_getter = getattr(val, '_channel_', None)
+    channel_getter = getattr(val, "_channel_", None)
     if channel_getter is not None:
         warnings.warn(
-            '_channel_ is deprecated and will be removed in cirq 0.13, rename to _kraus_',
+            "_channel_ is deprecated and will be removed in cirq 0.13, rename to _kraus_",
             DeprecationWarning,
         )
 
-    kraus_getter = getattr(val, '_kraus_', None)
-    kraus_result = NotImplemented if kraus_getter is None else kraus_getter()
-    if kraus_result is not NotImplemented:
-        return tuple(kraus_result)
+    channel_result = NotImplemented if channel_getter is None else channel_getter()
+    if channel_result is not NotImplemented:
+        return tuple(channel_result)
+
+    _, kraus_result = _strat_kraus_from_kraus(val)
+    if kraus_result is not None and kraus_result is not NotImplemented:
+        return kraus_result
 
-    mixture_getter = getattr(val, '_mixture_', None)
-    mixture_result = NotImplemented if mixture_getter is None else mixture_getter()
-    if mixture_result is not NotImplemented and mixture_result is not None:
+    mixture_result = mixture(val, None)
+    if mixture_result is not None and mixture_result is not NotImplemented:
         return tuple(np.sqrt(p) * u for p, u in mixture_result)
 
-    unitary_getter = getattr(val, '_unitary_', None)
-    unitary_result = NotImplemented if unitary_getter is None else unitary_getter()
-    if unitary_result is not NotImplemented and unitary_result is not None:
+    unitary_result = unitary(val, None)
+    if unitary_result is not None and unitary_result is not NotImplemented:
         return (unitary_result,)
 
-    channel_result = NotImplemented if channel_getter is None else channel_getter()
-    if channel_result is not NotImplemented:
-        return tuple(channel_result)
+    decomposed = decompose(val)
+
+    if decomposed != [val]:
+        qubits: List[Qid] = []
+        for x in decomposed:
+            qubits.extend(x.qubits)
+        qubits = sorted(list(set(qubits)))
+        limit = (4 ** np.prod(len(qubits))) ** 2
+
+        kraus_list = list(map(lambda x: _kraus_tensor(x, qubits, default), decomposed))
+        if not any([_check_equality(x, default) for x in kraus_list]):
+            kraus_result = kraus_list[0]
+            for i in range(1, len(kraus_list)):
+                kraus_result = [op_2.dot(op_1) for op_1 in kraus_result for op_2 in kraus_list[i]]
+                assert (
+                    len(kraus_result) < limit
+                ), f"{val} kraus decomposition had combinatorial explosion."
+            return tuple(kraus_result)
```

Comment:
Use `apply_channel` and `apply_mixture` instead of rolling out your own serial concatenation logic.